controllers/storage_test.go

```diff
@@ -259,3 +259,108 @@ func TestStorageRemoveAllButCurrent(t *testing.T) {
 		}
 	})
 }
+
+func TestStorageCopyFromPath(t *testing.T) {
+	type File struct {
+		Name    string
+		Content []byte
+	}
+
+	dir, err := createStoragePath()
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(cleanupStoragePath(dir))
+
+	storage, err := NewStorage(dir, "hostname", time.Minute)
+	if err != nil {
+		t.Fatalf("error while bootstrapping storage: %v", err)
+	}
+
+	createFile := func(file *File) (absPath string, err error) {
+		defer func() {
+			if err != nil && dir != "" {
+				os.RemoveAll(dir)
+			}
+		}()
+		dir, err = os.MkdirTemp("", "test-files-")
+		if err != nil {
+			return
+		}
+		absPath = filepath.Join(dir, file.Name)
+		if err = os.MkdirAll(filepath.Dir(absPath), 0755); err != nil {
+			return
+		}
+		f, err := os.Create(absPath)
+		if err != nil {
+			return "", fmt.Errorf("could not create file %q: %w", absPath, err)
+		}
+		if n, err := f.Write(file.Content); err != nil {
+			f.Close()
+			return "", fmt.Errorf("could not write %d bytes to file %q: %w", n, f.Name(), err)
+		}
+		f.Close()
+		return
+	}
+
+	matchFile := func(t *testing.T, storage *Storage, artifact sourcev1.Artifact, file *File, wantErr bool) {
+		c, err := os.ReadFile(storage.LocalPath(artifact))
+		if err != nil {
+			t.Fatalf("failed reading file: %v", err)
+		}
+		if (string(c) != string(file.Content)) != wantErr {
```

Comment:
This seems to make assumptions between a correlation of match/mismatch and an error being returned which may yield false positives/negatives.

The two assertions here could be broken apart for added clarity.

Reply:
I meant to say that mismatch is expected for a test case. The `wantErr` is a bad name and will update it to be more clear.